core/src/main/scala/io/qbeast/spark/index/SparkPlanAnalyzer.scala

```diff
@@ -0,0 +1,120 @@
+import org.apache.spark.sql.catalyst.plans.logical.Sort
+import org.apache.spark.sql.DataFrame
+
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+
+trait SparkPlanAnalyzer {
+
+  /**
+   * Builds a list of non-deterministic operations in the logical plan
+   *
+   * A list of Non-Deterministic operations:
+   *   - LocalLimit without Sort Operation
+   *   - Sample
+   *   - Filter with non-deterministic condition
+   *
+   * @param plan
+   *   the logical plan
+   * @return
+   */
+  private def isLogicalPlanDeterministic(plan: LogicalPlan): Boolean = {
+    // Recursively traverse the logical plan to find non-deterministic operations
+    val isCurrentOperationDeterministic = plan match {
+      case LocalLimit(_, _: Sort) => true // LocalLimit with Sort is deterministic
+      case LocalLimit(_, _) => false
+      case Sample(_, _, false, _, child) =>
+        child match {
+          case QbeastRelation(_, _) =>
+            true // Sample over QbeastRelation is deterministic
+          case _ => false
+        }
+      case Filter(condition, _) => condition.deterministic
+      case _ => true
+    }
+
+    val areChildOperationsDeterministic = plan.children.forall(isLogicalPlanDeterministic)
+
+    isCurrentOperationDeterministic && areChildOperationsDeterministic
+  }
+
+  /**
+   * Extracts the select expressions from the logical plan for the provided column name
+   *
+   * @param logicalPlan
+   *   the logical plan
+   * @param columnName
+   *   the column name
+   * @return
+   */
+  private[index] def collectSelectExpressions(
+      logicalPlan: LogicalPlan,
+      columnName: String): Seq[Expression] = {
+    logicalPlan match {
+      case Project(projectList, _) =>
+        projectList.collect {
+          case Alias(child, name) if name == columnName => child
+          case expr if expr.references.map(_.name).contains(columnName) => expr
+        }
+      case _ =>
+        logicalPlan.children.flatMap(child => collectSelectExpressions(child, columnName))
+    }
+  }
+
+  /**
+   * Checks if a column is deterministic in the logical plan
+   * @param logicalPlan
+   *   the logical plan
+   * @param columnName
+   *   the column name
+   * @return
+   */
+
+  private def isColumnDeterministic(logicalPlan: LogicalPlan, columnName: String): Boolean = {
+    val expressionSet = collectSelectExpressions(logicalPlan, columnName)
+    expressionSet.forall(_.deterministic)
+  }
+
+  /**
+   * Analyzes the DataFrame to determine if it's execution is safely deterministic for indexing
+   *
+   *   - The logical plan of the DataFrame is checked for determinism
+   *   - The columns to analyze (to index) are checked for determinism
+   *
+   * @param dataFrame
+   *   the DataFrame to analyze
+   * @param columnsToAnalyze
+   *   the columns to analyze for determinism
+   * @return
+   */
+  def analyzeDataFrameDeterminism(
+      dataFrame: DataFrame,
+      columnsToAnalyze: Seq[String]): Boolean = {
```

Comment:
Instead of using `columnsToAnalyze: Seq[String],` pass the `Reivison` object. Filter the `Transformers` that have to be deterministic and check the determinism of their columns. `areColumnsToAnalyzeDeterministic` should be `true` if there's no such `Transformer` in the `Revision.`

We probably don't need to create `shouldBeDeterministic` for the Transformer trait for now.

Use something like 
```scala
revision.columnTransformers.filter {
  case t: LinearTransformer => true
  case _ => false
}
  .exist(isColumnDeterministic(_.columnName))
```

Reply:
I was thinking about this, but I wanted to narrow the scope of the class to avoid including anything from Qbeast. See, this SparkPlanAnalyzer should only be in charge of Analyzing the SparkPlan with the internals of Spark.

Reply:
On the shouldBeDeterministic, I think it's a better approach to push the acknowledge of determinism to the Transformer itself. 
Otherwise, each time you build a new Transformer, you must remember each time to add it (or not) to this other part of the code.

Reply:
Consider renaming it to something like "isBounded" to better reflect the nature of the default Transformation it will create, i.e., will all values be mapped to `[0, 1]` regardless of the min/max or other parameters used to create the Transformation? Using `shouldBeDeterministic` ties the property too much to this issue.

Reply:
Mm, I like it. It can work for other constraints in the future.

Reply:
Have a silly question, though. What is bounded? the Transformer or the Transformation?

Reply:
They can be in the Transformations because we can access them after computing the revision changes.

As for how to define this property, the thing that concerns us is whether the `transform` method can or cannot output values outside the range [0, 1]. Then, a `Transformation` is bounded only when `transform` as a function has a domain of [0,1] at max.